docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+# Quality of Service in Kubernetes
+
+**Author**: Ananya Kumar (@AnanyaKumar)
+
+**Status**: Draft proposal; prototype in progress.
+
+*This document presents the design of quality of service tiers for pods in Kubernetes, describes use cases, and lays out milestones for its development.*
+
+## Motivation
+
+Kubernetes allocates resources to pods in a simple way. Users specify limits for containers. For example, a container can have a 1gb memory limit. The scheduler uses the limit to launch containers. For example, the scheduler will not place 5 containers with a 1gb memory limit into a machine with 4gb memory. Additionally, Kubelet does not allow containers to exceed specified limits. In the current implementation, there aren’t robust mechanisms in place to ensure that pods run reliably on an overloaded system and nodes are not very efficiently utilized.
+
+**Pods with limits specified are not very well protected from pods without limits specified**. If a pod without an explicit memory limit goes overboard and uses lots of memory, other pods (with specified memory limits) might be killed. This is bad, because users often want a way to launch containers that have resources guarantees, and that stay up reliably.
+
+On the other hand, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Pods often don’t use all the resources that they request which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a node with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in more pods onto the node, but Kubernetes will not schedule new pods on the node.
+
+This proposal provides mechanisms for solving both the reliability and efficiency problems, by allowing pods to specify levels of resource guarantees. Containers will be able to *request* for a minimum resource guarantee. The *request* is different from the *limit* - containers will not be allowed to exceed resource limits.
+
+Additionally, we will support 3 different tiers of pods. Guaranteed pods (where requests and limits are equal) will receive the highest priority and will get all the resources they need. Burstable pods, which specify request < limit will be guaranteed their request and can “burst” and use more resources when available. Best-Effort pods, which don’t specify a request, can use resources only if not being used by other pods. In Borg, supporting best-effort pods increased utilization by about 20%, and we hope to see similar improvements in Kubernetes.

```

Comment:
Configuring the ratio in LimitRange won't work, unfortunately. We'd need to explicitly record the decision in each pod, and we'd also need to think about how pods with different policies would interact. Probably better to set it via system configuration for now.

We should think about how to treat best-effort containers with non-zero requests, though, if we go with this approach.